api/src/services/files.ts

```diff
@@ -83,11 +90,19 @@ export class FilesService extends ItemsService {
 		}
 
 		const fileExtension =
-			path.extname(payload.filename_download!) || (payload.type && '.' + extension(payload.type)) || '';
+			path.extname(payload.filename_download!) || (payload?.type && '.' + extension(payload.type)) || '';
+
+		if (!path.extname(payload.filename_download!) && payload?.type) {
+			payload.filename_download = payload.filename_download + '.' + extension(payload.type);
+		}
 
 		// The filename_disk is the FINAL filename on disk
 		payload.filename_disk ||= primaryKey + (fileExtension || '');
 
+		if (existingFile && path.extname(payload.filename_disk!) !== fileExtension) {
```

Comment:
This replaces the `filename_disk` with an updated extension if replacing an existing file where the extension doesn't match.